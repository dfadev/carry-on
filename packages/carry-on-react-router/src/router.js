import { mutateSetA, mutateMerge, toPath, getInA } from "carry-on-utils";
import { createLocation, createBrowserHistory } from "history";
import matchPath from "./matchPath";

function isModifiedEvent(event) {
  return !!(event.metaKey || event.altKey || event.ctrlKey || event.shiftKey);
}

const router = (history = createBrowserHistory(), path = "app.history") => {
  let isPaused = false;
  const historyPath = toPath(path);

  // router state
  const state = ({ get, set }) => {
    const prevHist = getInA(get(), history);
    if (prevHist !== undefined && prevHist.unlisten) prevHist.unlisten();

    const unlisten =
      history.listen &&
      history.listen(
        () =>
          !isPaused &&
          set(s => {
            const hist = getInA(s, historyPath);
            mutateMerge(hist, history);
          }, "History Change")
      );

    // create an event handler for a link click
    const handleClick =
      ({ onClick, target, replace, to, force }) =>
      event => {
        if (onClick) onClick(event);

        if (
          !event.defaultPrevented && // onClick prevented default
          event.button === 0 && // ignore everything but left clicks
          (!target || target === "_self") && // let browser handle "target=_blank" etc.
          !isModifiedEvent(event) // ignore clicks with modifier keys
        ) {
          event.preventDefault();

          const method = replace ? history.replace : history.push;

          if (force) window.location = to;
          else method(to);
        }
      };

    // create an href based on the current location
    function getHref(to) {
      const hist = getInA(get(), historyPath);
      let location = to;
      if (typeof to === "string") {
        if (to.charAt(0) !== "/") {
          location = createLocation(to, null, null, hist.location);
        } else {
          location = createLocation(to);
        }
      }
      const href = location ? hist.createHref(location) : "";
      return href;
    }

    const stage = {
      unlisten,
      ...history,
      entries: history.entries ? history.entries.slice() : [],
      location: { ...history.location },

      match: {
        path: "/",
        url: "/",
        params: {},
        isExact: history.location && history.location.pathname === "/"
      },
      handleClick,
      getHref,
      matchPath: opts =>
        matchPath(getInA(get(), historyPath).location.pathname, opts)
    };

    if (history.staticContext) {
      const setContext = cs => {
        set(s => {
          const hist = getInA(s, historyPath);
          hist.staticContext = cs(hist);
        }, "History Change");
      };

      stage.push = stage.push(setContext);
      stage.replace = stage.replace(setContext);
    }

    return mutateSetA({}, historyPath, stage);
  };

  // router middleware
  const middleware =
    ({ get, next, isNested }) =>
    (action, opts) => {
      const nextState = next(action, opts);
      const type = opts && (opts.type || opts);
      if (isNested() || type !== "Time Travel") return nextState;

      // time travel should replace history location
      isPaused = true;
      const hist = getInA(get(), historyPath);
      if (
        window.location.pathname !== hist.location.pathname ||
        window.location.search !== hist.location.search ||
        window.location.hash !== hist.location.hash
      )
        hist.replace(hist.location);
      isPaused = false;

      return nextState;
    };

  return { state, middleware };
};

export default router;
