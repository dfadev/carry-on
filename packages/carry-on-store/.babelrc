{
  "presets": [
    ["env", {
      "modules": false,
      "targets": { "browsers": ["last 2 versions"] }
    }],
    "stage-2",
    "optimized"
  ],
  "env": {
    "development": {
      "plugins": ["external-helpers"],
    },
    "production": {
      "plugins": ["external-helpers"],
    },
    "test": {
      "presets": ["env", "stage-2"]
    }
  }
}
