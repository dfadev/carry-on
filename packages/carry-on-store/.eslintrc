{
  "parser": "babel-eslint",
  "extends": ["airbnb-base", "prettier"],
  "env": { "es6": true, "browser": true },
  "rules": {
    "no-empty-pattern": "off",
    "no-debugger": "warn",
    "no-unused-expressions": "off",
    "no-unused-vars": "warn",
    "no-plusplus": "off",
    "import/no-unresolved": "off",
    "no-restricted-syntax": "off",
    "no-underscore-dangle": "off",
    "prefer-template": "off",
    "spaced-comment": "off",
    "no-param-reassign": "off",
    "no-return-assign": "off",
    "no-multi-assign": "off",
    "one-var": "off",
    "prefer-destructuring": "off",
  }
}
