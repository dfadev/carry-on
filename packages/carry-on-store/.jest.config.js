module.exports = {
  //setupTestFrameworkScriptFile: "./node_modules/jest-enzyme/lib/index.js",
  setupTestFrameworkScriptFile: "./test/jest.setup.js",
  verbose: true,
  testPathIgnorePatterns: ["<rootDir>/build/", "<rootDir>/node_modules/", "<rootDir>/examples/", "<rootDir>/lib", "<rootDir>/src/jest.setup.js"],
  collectCoverage: true,
  coverageDirectory: "<rootDir>/coverage/",
  moduleNameMapper: {
    "^carry-on-utils$":
      "<rootDir>/node_modules/carry-on-utils/dist/carry-on-utils.esm.js",
  }
};
