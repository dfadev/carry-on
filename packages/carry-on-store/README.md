# carry-on-store

# STATUS: WIP SUBJECT TO CHANGE

- This is the core module that provides the store.

Read the documentation at: https://dfadev.github.io/carry-on
