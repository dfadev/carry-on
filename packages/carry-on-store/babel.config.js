module.exports = {
  presets: [
    [
      "@babel/preset-env",
      {
        modules: false,
        targets: {
          browsers: [">3%"]
        }
      }
    ]
  ],
  env: {
    test: {
      plugins: [
        "@babel/plugin-transform-modules-commonjs",
        [
          "@babel/plugin-transform-runtime",
          {
            regenerator: true
          }
        ]
      ],
      presets: ["@babel/preset-env"]
    }
  }
};
