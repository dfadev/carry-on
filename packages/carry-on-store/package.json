{
  "name": "carry-on-store",
  "version": "1.0.4",
  "description": "State manager",
  "source": "./src/index.js",
  "module": "./dist/carry-on-store.esm.js",
  "main": "./dist/carry-on-store.cjs.js",
  "browser": "./dist/carry-on-store.umd.js",
  "scripts": {
    "clean": "rimraf lib dist es coverage",
    "format": "prettier --write \"{src,test}/**/*.js\"",
    "format:check": "prettier --list-different \"{src,test}/**/*.js\"",
    "lint": "eslint src test",
    "pretest": "npm run build",
    "test": "cross-env BABEL_ENV=test jest --config ./.jest.config.js",
    "test:watch": "npm test -- --watchAll",
    "test:cov": "npm test -- --coverage",
    "build": "rollup -c rollup.config.js",
    "build-dev": "rollup -c rollup.config.js --configDebug",
    "watch": "npm run build -- -w"
  },
  "author": "Russ Panula <russ@dfa.net>",
  "license": "MIT",
  "devDependencies": {
    "@babel/core": "^7.12.10",
    "@babel/plugin-proposal-class-properties": "^7.12.1",
    "@babel/plugin-transform-modules-commonjs": "^7.12.1",
    "@babel/plugin-transform-runtime": "^7.12.10",
    "@babel/preset-env": "^7.12.11",
    "@testing-library/jest-dom": "^5.11.8",
    "babel-eslint": "^10.1.0",
    "babel-jest": "^26.6.3",
    "cross-env": "^7.0.3",
    "eslint": "^7.17.0",
    "eslint-config-airbnb-base": "^14.2.1",
    "eslint-config-prettier": "^7.1.0",
    "eslint-plugin-import": "^2.22.1",
    "jest": "^26.6.3",
    "prettier": "^2.2.1",
    "rimraf": "^3.0.2",
    "rollup": "^2.36.1",
    "rollup-plugin-babel": "^4.3.3",
    "rollup-plugin-eslint": "^7.0.0",
    "snapshot-diff": "^0.8.1"
  },
  "sideEffects": false,
  "gitHead": "e21942a1d6e23d964f9b74bd1c9c3744cb7b3c18",
  "dependencies": {
    "@babel/runtime": "^7.12.5",
    "carry-on-utils": "^1.0.4",
    "immer": "^8.0.0"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/dfadev/carry-on.git"
  }
}
