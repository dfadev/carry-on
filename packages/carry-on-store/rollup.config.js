/** @format */
import babel from "rollup-plugin-babel";
import { eslint } from "rollup-plugin-eslint";

const pkg = require("./package.json");
const input = pkg.source;
const output = [
  {
    file: pkg.module,
    format: "esm",
    name: pkg.name,
    sourcemap: true
  },
  {
    file: pkg.main,
    format: "cjs",
    name: pkg.name,
    sourcemap: true
  },
  {
    file: pkg.browser,
    format: "umd",
    name: pkg.name,
    sourcemap: true,
    globals: {
      immer: "immer",
      "carry-on-utils": "carryOnUtils"
    }
  }
];
const eslintPlugin = eslint();
const babelPlugin = babel();
const external = Object.keys(pkg.dependencies || {}).concat(
  Object.keys(pkg.peerDependencies || {})
);
const watch = {
  clearScreen: false,
  chokidar: {
    usePolling: true
  }
};
const defaultConfig = {
  plugins: [eslintPlugin, babelPlugin],
  input,
  output,
  external,
  watch
};

export default defaultConfig;
