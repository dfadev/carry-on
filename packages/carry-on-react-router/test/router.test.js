import { register, connect, initStores } from "carry-on-store";
import router from "../src/router";

it("router matches snapshot", () => {
  register(router());
  expect(connect()).toMatchSnapshot();
});
