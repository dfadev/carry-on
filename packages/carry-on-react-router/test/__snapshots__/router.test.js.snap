// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`router matches snapshot 1`] = `
Object {
  "app": Object {
    "history": Object {
      "action": "POP",
      "block": [Function],
      "createHref": [Function],
      "getHref": [Function],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "handleClick": [Function],
      "length": 1,
      "listen": [Function],
      "location": Object {
        "hash": "",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
      "match": Object {
        "isExact": true,
        "params": Object {},
        "path": "/",
        "url": "/",
      },
      "matchPath": [Function],
      "push": [Function],
      "replace": [Function],
      "unlisten": [Function],
    },
  },
}
`;
